Do not merge yet - 02 updates for review

saradickinson opened the pull request.

saradickinson committed:
First pass at further updates, mainly around padding. More to come on connection handling….

saradickinson committed:
Further updates to connection handling.

pallaviaras commented on draft-ietf-dprive-xfr-over-tls.md:
```diff
@@ -333,29 +336,186 @@ The figure below provides an outline of the AXoT mechanism including NOTIFYs.
+the connection open for a few seconds ([@RFC7766] recommends that servers use
+timeouts of at least a few seconds).
+
+An AXoT server MUST be able to handle multiple IXFR requests on a
```
should this be multiple AXFR requests?

saradickinson replied:
Yup :-)

pallaviaras commented on draft-ietf-dprive-xfr-over-tls.md:
```diff
@@ -333,29 +336,186 @@ The figure below provides an outline of the AXoT mechanism including NOTIFYs.
+For AXoT this requirement is restated: AXoT clients SHOULD re-use an existing
+open TLS connection when starting any new AXoT session to the same primary, and
+for issuing SOA queries, instead of opening a new connection. (Non-AXoT session
+traffic can also use an open TLS connection.) The number of TLS connections
```
Instead of "can also use" should we say RECOMMENDED to use?

ShivanKaul replied:
I guess we don't want to mandate what non-XoT traffic does with the existing TLS connection?

ShivanKaul replied:
Although we _are_ saying that the absolute number of TLS connections b/w secondary and primary should be minimized

saradickinson replied:
Thanks for these comments - this made me realise a couple of things. Firstly there is currently no RFC for DoT from recursive to authoritative, only for stub to resolver so in theory at this point there shouldn't be any other traffic for that case! Of course a recursive can also host a zone and anything can request a zone transfer of that zone but I don't think we can mandate this and should perhaps back it out a bit. Also, looking back at RFC7766 it says

"To mitigate the risk of unintentional server overload, DNS clients
   MUST take care to minimize the number of concurrent TCP connections
   made to any individual server.  It is RECOMMENDED that for any given
   client/server interaction there SHOULD be no more than one connection
   for regular queries, one for zone transfers, and one for each
   protocol that is being used on top of TCP (for example, if the
   resolver was using TLS).  However, it is noted that certain primary/
   secondary configurations with many busy zones might need to use more
   than one TCP connection for zone transfers for operational reasons
   (for example, to support concurrent transfers of multiple zones)."

which could be read as conflicting with RFC5936 except that doesn't use normative language...!  Let me think about how exactly we should update the text...

pallaviaras commented on draft-ietf-dprive-xfr-over-tls.md:
```diff
@@ -333,29 +336,186 @@ The figure below provides an outline of the AXoT mechanism including NOTIFYs.
+
+QUESTION FOR US: 
+
+* Not sure if we should dump all this section in a placeholder draft on padding
```
A placeholder draft to address "padding" is a good idea.

pallaviaras commented on draft-ietf-dprive-xfr-over-tls.md:
```diff
@@ -333,29 +336,186 @@ The figure below provides an outline of the AXoT mechanism including NOTIFYs.
+  about the zone contents to be inferred by watching different message sizes
+  over time if they are not all padded to the same size
+
+The goal of padding AXFR responses would be two fold:
```
We can add this section in the XOT draft to state reason for padding but the actual mechanism can be handled in a separate draft.

ShivanKaul replied:
+1

pallaviaras commented on draft-ietf-dprive-xfr-over-tls.md:
```diff
@@ -364,37 +524,128 @@ The figure below provides an outline of the IXoT mechanism including NOTIFYs.
-QUESTION: Since this is a new specification should there be a requirement that IXoT servers are RECOMMENDED to condense responses as described in Section 6 of [RFC1995]. [RFC1995] document says this is optional and MAY be done but it can significantly reduce the size of responses and may have implications for padding?
+Where any behavior is not explicitly described here, the behviour specified in
+[@RFC1995] MUST be followed. Any behaviour specified here takes precedence for
+AXoT implementations over that in [@RFC1995].
```
"AXoT implementations", should this be "IXoT implementations"

saradickinson replied:
yup again...

pallaviaras commented on draft-ietf-dprive-xfr-over-tls.md:
```diff
@@ -364,37 +524,128 @@ The figure below provides an outline of the IXoT mechanism including NOTIFYs.
+[@RFC1995] MUST be followed. Any behaviour specified here takes precedence for
+AXoT implementations over that in [@RFC1995].
+
+### IXoT connection and message handling
```
For AXoT and IXoT, the "connection and message handling" sections have a lot in common. Should we consolidate these sections?

saradickinson replied:
I do agree they are very similar but I wanted to keep them separate because the do relate to earlier specifications of IAXFR and IXFR in different ways and I suspect we might have to add some additional statements around handling AXFR sessions that are aborted..... I will try to reference the AXoT text from the IXoT section where possible though. I just thought of an AXoT specific issue with EDNS Keepalive that needs to be mentioned though

pallaviaras commented on draft-ietf-dprive-xfr-over-tls.md:
```diff
@@ -405,20 +656,52 @@ zone deltas and how many may be stored at any one time.
+transfers or the SOA of the primary being an unacceptable lag behind the SOA of
+an alternative primary.
+
+The other extreme can be considered a 'parallel primary connection' model. Here
```
should the draft say that the "parallel primary connection" is a RECOMMENDED model?

ShivanKaul replied:
We previously had "The mechanics of co-ordinating or enforcing such policies are out of the scope of this document but may be the subject of future operational guidance." in this section. It does seem like this is operational guidance affected by XoT but many other factors as well.

saradickinson replied:
Agree that I don't think there is a one size fits all answer here and it is an operational, not a protocol question. For now I've added this text: "Recommendation of a particular scheme is out of scope of this document but
implementations are encouraged to provide configuration options that allow
operators to make choices about this behaviour."  WDYT?

pallaviaras commented on draft-ietf-dprive-xfr-over-tls.md:
```diff
@@ -405,20 +656,52 @@ zone deltas and how many may be stored at any one time.
 
 ## TSIG
 
 TSIG [@RFC2845] provides a mechanism for two parties to exchange secret keys
```
The RFC2845 says "No provision has been made here for distributing the shared secrets;", so should the above line "provides a mechanism for two parties to exchange secret keys" be changed to "provides a mechanism to use secret keys shared amongst two parties"

saradickinson replied:
Good point - I've actually updated it to say "provides a mechanism for two or more parties to use shared
secret keys which can "...

pallaviaras commented on draft-ietf-dprive-xfr-over-tls.md:
```diff
@@ -405,20 +656,52 @@ zone deltas and how many may be stored at any one time.
 ## SIG(0)
 
-TBD
+SIG(0) [@RFC2535] similarly also provides a mechanism to digitally a sign DNS
```
Typo fix: "digitally sign a DNS message"

saradickinson replied:
thanks -fixed

ShivanKaul commented on draft-ietf-dprive-xfr-over-tls.md:
```diff
@@ -333,29 +336,186 @@ The figure below provides an outline of the AXoT mechanism including NOTIFYs.
+### Coverage and relationship to RFC5936
+
+[@RFC5936] re-specified AXFR providing additional guidance beyond that provided
+in [@RFC1045] and [@RFC1035]. For example, sections 4.1, 4.1.1 and 4.1.2 of
```
RFC1045 -> RFC1034

saradickinson replied:
thanks - fixed

ShivanKaul commented on draft-ietf-dprive-xfr-over-tls.md:
```diff
@@ -275,14 +274,16 @@ default in their latest releases. For BIND TCP connections are sometimes used
 [@RFC7766] to optimise the use of TCP connections and SHOULD use [@!RFC7858] to
 manage persistent connections.
 
+QUESTION FOR US: We don't go as far here as updating IXFR-over-TCP to match the behaviour outlined in section 5.4.2 for IXoT. Should we? 
```
We probably don't want to increase the scope of this already expanding doc more than we need to.

saradickinson replied:
I agree - just wanted to pose the question....

ShivanKaul commented on draft-ietf-dprive-xfr-over-tls.md:
```diff
@@ -275,14 +274,16 @@ default in their latest releases. For BIND TCP connections are sometimes used
 an IXFR is completed.
  
-<!--TODO: Look at packet captures from NSD and Knot Auth to see what they do.-->
+QUESTION FOR US: Look at packet captures from NSD and Knot Auth to see what they do.
```
I can take a look at this!

ShivanKaul commented on draft-ietf-dprive-xfr-over-tls.md:
```diff
@@ -333,29 +336,186 @@ The figure below provides an outline of the AXoT mechanism including NOTIFYs.
+compatibility since it is expected that all implementations of AXoT fully
+implement the behaviour described here.
+
+Where any behavior is not explicitly described here, the behviour specified in
```
nit: behviour -> behavior (we seem to be using American English in the doc and I just learnt that either American or British is fine [as long as the doc is internally consistent](https://tools.ietf.org/id/draft-flanagan-7322bis-03.html))

saradickinson replied:
My bad - I forget to switch my spell check to US English... I've been through and changed everything to use US English.

ShivanKaul commented on draft-ietf-dprive-xfr-over-tls.md:
```diff
@@ -333,29 +336,186 @@ The figure below provides an outline of the AXoT mechanism including NOTIFYs.
+    open connection.)"
+
+For AXoT this requirement is restated: AXoT clients SHOULD re-use an existing
+open TLS connection when starting any new AXoT session to the same primary, and
```
AXoT session -> AXoT transfers?

saradickinson replied:
I used session because that is what is used in RFC5936 and it think it is helpful to clearly distinguish an AXoT session (that is one request and potentially many responses) from an IXoT transaction or transfer which is just a query/response.  RFC5936 deals with some of the special error cases around AXFR due to the 'session' nature of the exchange which I think we also need to think about a bit more.....

ShivanKaul commented on draft-ietf-dprive-xfr-over-tls.md:
```diff
@@ -333,29 +336,186 @@ The figure below provides an outline of the AXoT mechanism including NOTIFYs.
+    single RR.  To interoperate with such clients, the server MAY
+    restrict response messages to a single RR."
+
+This is as opposed to the normal behaviour of containing a sufficient number of
```
nit: -"as"

saradickinson replied:
fixed

ShivanKaul commented on draft-ietf-dprive-xfr-over-tls.md:
```diff
@@ -364,37 +524,128 @@ The figure below provides an outline of the IXoT mechanism including NOTIFYs.
-connection. AN IXoT server MAY respond to those requests out of order.
-
-QUESTION: Since this is a new specification should there be a requirement that IXoT servers are RECOMMENDED to condense responses as described in Section 6 of [RFC1995]. [RFC1995] document says this is optional and MAY be done but it can significantly reduce the size of responses and may have implications for padding?
+Where any behavior is not explicitly described here, the behviour specified in
```
nit: "behviour"

ShivanKaul commented on draft-ietf-dprive-xfr-over-tls.md:
```diff
@@ -364,37 +524,128 @@ The figure below provides an outline of the IXoT mechanism including NOTIFYs.
+
+QUESTION: Since this is a new specification should there be a requirement that
+IXoT servers are RECOMMENDED to condense responses as described in Section 6 of
+[RFC1995]. [RFC1995] document says this is optional and MAY be done. Whilst it
```
Do we know if responses are usually condensed? We should try to limit barrier to adoption

saradickinson replied:
BIND does not condense, need to test other implementations. I doubt we can RECOMMEND this for the reason you give but it would be nice to push for it more...

ShivanKaul commented on draft-ietf-dprive-xfr-over-tls.md:
```diff
@@ -598,12 +855,15 @@ TBD
+
+* the risk that some level of zone activity might be inferred by observing zone transfer sizes and timing on encrypted connections (even with padding applied), in combination with obtaining SOA records by directly querying authoritative servers.
+* the risk that hidden primaries might be inferred or identified via observation of encrypted connections.
+* the risk of zone contents being obtained via zone enumeration techniques.
```
Do we think it's worthwhile here to point to the DNS encrypted traffic analysis research by Carmela T et al?

saradickinson replied:
Perhaps but I think all the work done to date is for stub to recursive traffic?

ShivanKaul commented on draft-ietf-dprive-xfr-over-tls.md:
```diff
@@ -333,29 +336,186 @@ The figure below provides an outline of the AXoT mechanism including NOTIFYs.
+implement the behaviour described here.
+
+Where any behavior is not explicitly described here, the behviour specified in
+[@RFC5953] MUST be followed. Any behaviour specified here takes precedence for
```
RFC 5953 -> RFC 5936?

saradickinson replied:
fixed

saradickinson committed:
Fix typos
Change spelling to US English
Update EDNS0 test for AXoT
Reference AXoT section from IXoT section to reduce duplicated text
Add text on multi-primary config options

saradickinson committed:
Remove all detailed text around padding, just leave minimal discussion and requirements.

saradickinson committed:
Add Shivan as author - switch Han to a contributor

saradickinson committed:
- A brain dump of issues around shared or dedicated connection use for DoT vs XoT.  Proposal is not fully coherent yet - needs work.
- Clean up of open questions
